keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  serviceAccountName: [3]
+  accessSecretName: [4]
+  bucketRequestName: [5]
+  bucketAccessClassName: [6]
+  bucketAccessName: [7]
+status:
+  conditions: [8]
+    - Type: {AccessGranted} [9]
+      Status: [10]
+      Reason: [11]
+      Message: [12]
+```
+
+1. `labels`: added by the controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by the controller to defer `BucketAccessRequest` deletion until backend deletion ops succeed.
+1. `serviceAccountName`: (optional) the name of a Kubernetes ServiceAccount in the same namespace.  This field is included to support cloud provider identity integration.  Should not be set when specifying `accessSecretName`.
+1. `accessSecretName`: (optional) the name of a Kubernetes Secret in the same namespace.  This field is used when there is not cloud provider identity integration.  Should not be set when specifying `serviceAccountName`.
+1. `bucketRequestName`: the name of the `BucketRequest` associated with this access request. From the bucket request, COSI knows the `Bucket` instance and thus bucket and its properties.
+1. `bucketAccessClassName`: name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance.
+1. `conditions`: the latest available observation of this `BucketAccessRequest`. (Phase is not used.)
+1. `Type`: only one type is supported: "AccessGranted". If the condition's `Status` is true then access to the bucket has been granted for the `Reason` provided. If `Status` is false then the bucket is unable to be accessed and, again, `Reason` states why.
+1. `Status`: one of true, false or unknown. If not specified then unknown is assumed. A true status indicates the this condition is true at the time of observation.
+1. `Reason`: one of "Bound", "Granting", "Revoking".
+1. `Message`: optional human readable message containing more details.
+
+#### BucketAccess
+
+A cluster-scoped administrative custom resource which encapsulates fields from the `BucketAccessRequest` (BAR) and the `BucketAccessClass` (BAC).  The purpose of the `BucketAccess` (BA) is to serve as communication path between provisioners and the central COSI controller.  In greenfield, the COSI controller creates `BucketAccess` instances for new `BucketAccessRequest`s. In brownfield the admin is expected to manually create the BA. There is one `BucketAccess` instance per `BucketAccessRequest`.
```

Comment:
> There is one `BucketAccess` instance per `BucketAccessRequest`.

What wasn't immediately clear to me from this sentence, but *is* from the schema of the object, is that a single BA is also tied to a single BAR. The sentence above may be interpreted as "A BAR has one BA, but a BA can be tied to multiple BARs", which is not true.

Given that, I wonder why BAs must be cluster-scoped?

Reply:
We did not want credentials to live in the application namespace.